File: jmolecules-axon/02-onion-cqrs-es-bank/src/main/java/org/jmolecules/example/axonframework/domain/command/model/MoneyTransferSaga.java
package org.jmolecules.example.axonframework.domain.command.model;

import org.axonframework.commandhandling.gateway.CommandGateway;
import org.axonframework.eventhandling.gateway.EventGateway;
import org.axonframework.modelling.saga.EndSaga;
import org.axonframework.modelling.saga.SagaEventHandler;
import org.axonframework.modelling.saga.StartSaga;
import org.axonframework.spring.stereotype.Saga;
import org.jmolecules.example.axonframework.domain.api.command.transfer.CancelMoneyTransferCommand;
import org.jmolecules.example.axonframework.domain.api.command.transfer.CompleteMoneyTransferCommand;
import org.jmolecules.example.axonframework.domain.api.command.transfer.ReceiveMoneyTransferCommand;
import org.jmolecules.example.axonframework.domain.event.transfer.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Saga
public class MoneyTransferSaga {

  private static final Logger logger = LoggerFactory.getLogger(MoneyTransferSaga.class);

  private String moneyTransferId;
  private String sourceAccountId;
  private String targetAccountId;
  private int amount;

  @SagaEventHandler(associationProperty = "moneyTransferId")
  @StartSaga
  public void on(MoneyTransferRequestedEvent evt, CommandGateway commandGateway) {
    this.moneyTransferId = evt.moneyTransferId();
    this.sourceAccountId = evt.sourceAccountId();
    this.targetAccountId = evt.targetAccountId();
    this.amount = evt.amount();

    logger.info("MoneyTransfer [{}] started with: {}", this.moneyTransferId, evt);

    var receiveCmd = new ReceiveMoneyTransferCommand(this.targetAccountId, this.moneyTransferId, this.amount);
    commandGateway.send(receiveCmd, (commandMessage, commandResultMessage) -> {
      if (commandResultMessage.isExceptional()) {
        // on error cancel
        var cancelCmd = new CancelMoneyTransferCommand(
          this.sourceAccountId,
          this.moneyTransferId,
          commandResultMessage.exceptionResult().getMessage()
        );
        commandGateway.sendAndWait(cancelCmd);
      }
    });
  }

  @SagaEventHandler(associationProperty = "moneyTransferId")
  public void on(MoneyTransferReceivedEvent evt, CommandGateway commandGateway) {
    logger.trace("on({})", evt);
    var completeCmd = new CompleteMoneyTransferCommand(sourceAccountId, moneyTransferId);

    // applied simplification: completion can/will not fail. Otherwise, we would have to implement compensation on the target account (2-phase commit).
    commandGateway.sendAndWait(completeCmd);
  }

  @SagaEventHandler(associationProperty = "moneyTransferId")
  @EndSaga
  public void on(MoneyTransferCompletedEvent evt, EventGateway eventGateway) {
    logger.trace("on({})", evt);
    // Milestone event
    var transferredEvent = new MoneyTransferredEvent(
      moneyTransferId,
      sourceAccountId,
      targetAccountId,
      amount
    );
    eventGateway.publish(transferredEvent);
    logger.info("MoneyTransfer [{}] finished with: {}", this.moneyTransferId, transferredEvent);
  }

  @Override
  public String toString() {
    return "MoneyTransferSaga{" +
      "moneyTransferId='" + moneyTransferId + '\'' +
      ", sourceAccountId='" + sourceAccountId + '\'' +
      ", targetAccountId='" + targetAccountId + '\'' +
      ", amount=" + amount +
      '}';
  }
}

File: jmolecules-axon/03-onion-axon-cqrs-es-bank/src/main/kotlin/org/jmolecules/example/axonframework/infrastructure/adapter/policy/MoneyTransferSaga.kt
package org.jmolecules.example.axonframework.infrastructure.adapter.policy

import mu.KLogging
import org.axonframework.commandhandling.CommandResultMessage
import org.axonframework.commandhandling.gateway.CommandGateway
import org.axonframework.eventhandling.gateway.EventGateway
import org.axonframework.modelling.saga.SagaEventHandler
import org.axonframework.modelling.saga.SagaLifecycle
import org.axonframework.modelling.saga.StartSaga
import org.axonframework.spring.stereotype.Saga
import org.jmolecules.event.annotation.DomainEventHandler
import org.jmolecules.event.annotation.DomainEventPublisher
import org.jmolecules.example.axonframework.core.model.moneytransfer.MoneyTransfer
import org.jmolecules.example.axonframework.core.model.moneytransfer.Reason
import org.jmolecules.example.axonframework.core.model.moneytransfer.command.CancelMoneyTransferCommand
import org.jmolecules.example.axonframework.core.model.moneytransfer.command.CompleteMoneyTransferCommand
import org.jmolecules.example.axonframework.core.model.moneytransfer.command.ReceiveMoneyTransferCommand
import org.jmolecules.example.axonframework.core.model.moneytransfer.event.*

@Saga
class MoneyTransferSaga {

  companion object : KLogging()

  private lateinit var moneyTransfer: MoneyTransfer

  @StartSaga
  @SagaEventHandler(associationProperty = "moneyTransferId")
  @DomainEventHandler(namespace = "axon.bank", name = "MoneyTransferRequestedEvent")
  fun on(evt: MoneyTransferRequestedEvent, commandGateway: CommandGateway) {

    logger.info { "MoneyTransfer [${evt.moneyTransferId}] started with: $evt" }

    moneyTransfer = MoneyTransfer(
      moneyTransferId = evt.moneyTransferId,
      sourceAccountId = evt.sourceAccountId,
      targetAccountId = evt.targetAccountId,
      amount = evt.amount
    )

    val receiveCmd = ReceiveMoneyTransferCommand(
      moneyTransfer.targetAccountId,
      moneyTransfer.moneyTransferId,
      moneyTransfer.amount
    )
    commandGateway.send<ReceiveMoneyTransferCommand, Any>(receiveCmd) { _, commandResultMessage: CommandResultMessage<*> ->
      if (commandResultMessage.isExceptional) {
        // on error cancel
        val cancelCmd = CancelMoneyTransferCommand(
          moneyTransferId = moneyTransfer.moneyTransferId,
          sourceAccountId = moneyTransfer.sourceAccountId,
          targetAccountId = moneyTransfer.targetAccountId,
          reason = Reason.reasonForMessage(commandResultMessage.exceptionResult().message)
        )
        commandGateway.sendAndWait<Any>(cancelCmd)
      }
    }
  }

  @DomainEventHandler(namespace = "axon.bank", name = "MoneyTransferReceivedEvent")
  @SagaEventHandler(associationProperty = "moneyTransferId")
  fun on(evt: MoneyTransferReceivedEvent, commandGateway: CommandGateway) {
    logger.trace { "on($evt)" }
    val completeCmd = CompleteMoneyTransferCommand(
      sourceAccountId = moneyTransfer.sourceAccountId,
      moneyTransferId = moneyTransfer.moneyTransferId
    )
    commandGateway.send<CompleteMoneyTransferCommand, Any>(completeCmd) { _, commandResultMessage: CommandResultMessage<*> ->
      if (commandResultMessage.isExceptional) {
        // on error cancel
        val cancelCmd = CancelMoneyTransferCommand(
          moneyTransferId = moneyTransfer.moneyTransferId,
          sourceAccountId = moneyTransfer.sourceAccountId,
          targetAccountId = moneyTransfer.targetAccountId,
          reason = Reason.reasonForMessage(commandResultMessage.exceptionResult().message)
        )
        commandGateway.sendAndWait<Any>(cancelCmd)
      }
    }
  }

  @SagaEventHandler(associationProperty = "moneyTransferId")
  @DomainEventHandler(namespace = "axon.bank", name = "MoneyTransferCompletedEvent")
  @DomainEventPublisher(
    publishes = "axon.bank.MoneyTransferredEvent",
    type = DomainEventPublisher.PublisherType.EXTERNAL
  )
  fun on(evt: MoneyTransferCompletedEvent, eventGateway: EventGateway) {
    logger.trace { "on($evt)" }
    // Milestone event
    val transferredEvent = MoneyTransferredEvent(
      moneyTransferId = moneyTransfer.moneyTransferId,
      sourceAccountId = moneyTransfer.sourceAccountId,
      targetAccountId = moneyTransfer.targetAccountId,
      amount = moneyTransfer.amount
    )
    eventGateway.publish(transferredEvent)
    logger.info { "MoneyTransfer [${moneyTransfer.moneyTransferId}] finished with: $transferredEvent" }

    SagaLifecycle.end()
  }

  @SagaEventHandler(associationProperty = "moneyTransferId")
  @DomainEventHandler(namespace = "axon.bank", name = "MoneyTransferCancelledEvent")
  fun on(evt: MoneyTransferCancelledEvent, eventGateway: EventGateway) {
    logger.trace { "on($evt)" }
    logger.info { "MoneyTransfer [${moneyTransfer.moneyTransferId}] cancelled with" }
    SagaLifecycle.end()
  }

  override fun toString(): String {
    return moneyTransfer.toString()
  }
}
